src/models/decomposition_models.jl

```diff
@@ -362,21 +318,47 @@ function MMI.fit(model::FactorAnalysis, verbosity::Int, X)
     return fitresult, cache, report
 end
 
-function MMI.fitted_params(::FactorAnalysis, fr)
-    return (projection=MS.projection(fr),)
-end
-
-function MMI.transform(::FactorAnalysis, fr::FactorAnalysisResultType, X)
-    # X is n x d, need to transpose twice
-    Xarray = MMI.matrix(X)
-    Xnew = transpose(MS.transform(fr, transpose(Xarray)))
-    return MMI.table(Xnew, prototype=X)
-end
-
 metadata_model(FactorAnalysis,
     input=Table(Continuous),
     output=Table(Continuous),
     weights=false,
     descr=FactorAnalysis_DESCR,
     path="$(PKG).FactorAnalysis"
 )
+
+####
+#### Common interface
+####
+
+model_types = [
+    (PCA, PCAFitResultType),
+    (KernelPCA, KernelPCAFitResultType),
+    (ICA, ICAFitResultType),
+    (PPCA, PPCAFitResultType),
+    (FactorAnalysis, FactorAnalysisResultType)
+]
+
+for (M, MFitResultType) in model_types
+    println("Defining methods for $M")
+
+    @eval function MMI.fitted_params(::$M, fr)
+        return (projection=copy(MS.projection(fr)),)
+    end
+
+    @eval function MMI.transform(::$M, fr::$MFitResultType, X)
+        # X is n x d, need to transpose twice
+        Xarray = MMI.matrix(X)
+        Xnew = transpose(MS.transform(fr, transpose(Xarray)))
+        return MMI.table(Xnew, prototype=X)
+    end
+
+    if hasmethod(MS.reconstruct, Tuple{MFitResultType{Float64}, Matrix{Float64}})
```

Comment:
```suggestion
    if hasmethod(MS.reconstruct, Tuple{MFitResultType{T}, AbstractMatrix{T}} where {T} )
```
not really a big deal but could you replace with above.